### DISCUSSION TITLE
代码块中“等宽英文字体”的自定义与实现方式

### DISCUSSION THREAD

**Discussion by F9y4ng** [F9y4ng/GreasyFork-Scripts#74] [经验分享(Share Experience)]
## 不定期更新等宽字体的应用<a name="rules"></a> 
> [!TIP]
> 请在以下列表中选择你需要的站点添加至[这里](#monosite)，如果您可提供新的站点定义请回复本贴，我会定期更新至列表。
```txt
@aistudio.google.com##.inline-code
@bing.com##.devmag_code *,.rd_inl_code,.rd_inl_code *
@latexlive.com###div_copy,#div_copy *,textarea
@bgm.tv|bangumi.tv|chii.in##textarea#formScript,textarea#formStyle
@github.com##[class^='ReferencedEvent-module__Box'] a p,#copilot-chat-textarea-preview,.text-mono
@zh.wikipedia.org##.ve-ce-paragraphNode
@请替换为AdguardHome访问域名或IP（须含端口号）##.font-monospace *,textarea
@pastebin.com##.highlighted-code *
@.notion.so##div.notion-code-block *
@bbs.kafan.cn##div[id^='code_'] *
@www.wolai.com##div[data-block-type='code'] *
@greasyfork.org##.diff *,#script_version_code
@www.w3schools.com##textarea,.w3-code *
@www.jb51.net##body div .syntaxhighlighter *
@www.cnblogs.com##.syntaxhighlighter *
@www.runoob.com##.example_code,.example_code *
@www.yuque.com##.ne-code,.ne-code *,ne-code,ne-code *
@ost.51cto.com###result [class*=language-],.prettyprint *,div[class*=language-] *
@github.io##.line-numbers *
```

# 自定义等宽字体工具介绍

在访问一些技术型网站时，通常会遇到`<pre>, <code>`包括的代码块，网站通常会使用代码着色、ESlint代码检查等等高级功能，对于字体而言，每个网站均有不同。为了使用户可以统一等宽字体的渲染，自定义等宽字体工具诞生了。

- 字体渲染（自用脚本）内置了针对等宽英文字体的个性化渲染设置。在默认情况下，等宽字体渲染功能时关闭的，你需要通过打开自定义等宽字体工具开启该功能。

  如果您需要自定义等宽字体的渲染，请勿轻易删除**排除渲染的HTML标签**中的`,pre,pre *,code,code *`

> [!NOTE]
>   1. 开启自定义等宽字体控制面板
>   <img src="https://github.com/user-attachments/assets/8bf7a67e-9e31-40d9-bf53-4ac10ad6b9e8" />  <img  src="https://github.com/user-attachments/assets/cc42a817-d368-47a4-8270-4ff5955ed98b" />
>   2. 在**自定义等宽字体**设置界面关闭开关即可<a name="monosite"></a>
>   <img width="300" src="https://github.com/user-attachments/assets/44c3332d-ff56-4887-8bf1-829545b87210" /> <img width="300" src="https://github.com/user-attachments/assets/bcab096c-4272-4f9c-8553-eaa69c80efb6" />
> 

- 开启等宽字体渲染后，在用户未自定义等宽字体时，脚本将默认使用脚本内置的等宽字体规则。当用户按填写要求提供自定义等宽英文字体时，优先使用用户定义的字体。**注意：此时中文字体将直接调用你设定的中文字体。**

- 自定义根域及样式选择器匹配规则，每行只能允许一组规则，相同站点不同规则可重复添加，相同样式选择器可定义多个根域名进行匹配。

- 等宽字体针对OpenType字体提供了高级属性font-feature-setting的设置功能。

- 等宽字体应用的根域及元素规则 （不定期更新，自取自用）
在日常使用时，并非全部网站都使用`<pre>`或`<code>`来定义代码块，部分网站使用其他的元素来定义代码块时，默认的规则将失效。为解决此问题，脚本新增了针对特殊网站的设置：[点这里查看](#rules)常用代码型网站的兼容代码。

# 自定义等宽字体的下载
点击[这里](https://github.com/F9y4ng/GreasyFork-Scripts/discussions/46#discussioncomment-2581520)，获取推荐的英文等宽字体。

**Comment by aboutZZ** [comment]


**Reply by F9y4ng** [reply]
似乎是正常的，如图：
![test](https://github.com/F9y4ng/GreasyFork-Scripts/assets/15855068/31cf66bd-f298-4612-abca-7c7e906fc207)

## 新版v2023.05.06.1更新后增加了自定义等宽字体的开关，需要手动去开启。
![开关](https://user-images.githubusercontent.com/15855068/236817716-0fa79f74-a025-42aa-ad58-b10c7a269341.png)

**Reply by aboutZZ** [reply]
按照你提供的规则加上去 好了

**Comment by Xia0xia0Xia0** [comment]
我推荐一个卡饭的代码块：
```CSS
@bbs.kafan.cn##div[id^='code_'] *
```

**Reply by F9y4ng** [reply]
感谢分享😀

**Comment by popokcn** [comment]
分享一个宝塔面板中在线编辑器的规则
@example.com###ace_editor

不知道前面域名可不可以模糊查找，应该用了ace_editor这个编辑器的，都需要等宽吧

**Reply by F9y4ng** [reply]
等宽字体默认规则已包含该编辑器，无需重复添加，感谢你的建议。

**Comment by wshizhx** [comment]
请添加一下飞牛OS，docker 查看日志的，也是不能正常显示
代码是这一段    div dir="ltr" class="terminal xterm xterm-dom-renderer-owner-1"

**Reply by F9y4ng** [reply]
> 请添加一下飞牛OS，docker 查看日志的，也是不能正常显示
> 代码是这一段    div dir="ltr" class="terminal xterm xterm-dom-renderer-owner-1"

没有本地环境只能靠猜了，添加以下代码试试看，其中URL替换为`你的本地访问地址`，如有端口号也需要加上，不包含`http(s)://`。

```
@URL##.terminal *
```

**Comment by nuu987** [comment]
```
@pastebin.com##.highlighted-code *
```
pastebin is a text content hosting provider. this fix would apply the monospace style correctly.

**Reply by F9y4ng** [reply]
Thanks, I've added it to the list.

**Comment by eachann1024** [comment]
不能 code 标签都加上吗？

**Reply by F9y4ng** [reply]
what are you talking about?

**Comment by skt-shinyruo** [comment]
太有用了👍

**Comment by peaynan** [comment]
推荐一个bangumi超合金组件（js脚本）脚本区和样式区文本的规则

```
@bgm.tv##textarea#formScript, textarea#formStyle
@bangumi.tv##textarea#formScript, textarea#formStyle
@chii.in##textarea#formScript, textarea#formStyle
```

三个不同的域名下的

**Reply by F9y4ng** [reply]
感谢，已新增至列表。这三个域名指向的是同一个站点吧？

**Reply by peaynan** [reply]
是同一站点

**Comment by LouieMSD** [comment]
```
@aistudio.google.com##.inline-code
```
aistudio.google.com 的 inline-code

**Reply by F9y4ng** [reply]
感谢，已新增至列表。

**Comment by bbqjam1145** [comment]
这边想问一下，我已经把浏览器和洛谷的默认字体改成了`Maple Mono`，为什么字体族还是`JetBrains Mono`？  
虽然可以直接在脚本里的自定义字体改，但我还是想问一下（

**Reply by F9y4ng** [reply]
Chrome会优先使用网页样式样式css中定义的字体进行渲染，如果未定义具体字体而是用通用字体族（如serif或serif-sans），或是定义了系统不存在的字体，才会采用浏览器中定义的字体作为fallback.

**Reply by bbqjam1145** [reply]
感谢大佬解惑，嘿嘿（
还有抱歉没有提供截图（
